### ISSUE TITLE
20.03 -> 20.09 regression: unable to deploy new (20.09) configuration due to initrd secrets of previous (20.03) configuration

### ISSUE THREAD

**Issue by wizeman** [NixOS/nixpkgs#101462]
**Describe the bug**
When upgrading a server from 20.03 to 20.09 (using nixops), I am unable to deploy the configuration. Instead, the following occurs:

```
homeserv> copying closure...                                                                   
wizynet> closures copied successfully                                                                                                                                                         
homeserv> updating GRUB 2 menu...                                                                                                                                                             
homeserv> 3 blocks                                                                             
homeserv> 3 blocks                                                                             
homeserv> cp: cannot stat '/backup/root/src/wizynet/secrets/dropbear_host_rsa_key': No such file or directory                                                                                 
homeserv> failed to create initrd secrets: No such file or directory                                                                                                                          
```

Note that the `dropbear_host_rsa_key` file doesn't exist anywhere in the configuration I'm deploying!
In fact, that file has never existed on the target machine, it is an artifact of the previous configuration!

You can see this in this configuration change that I was forced to do when upgrading from 20.03 -> 20.09:

```diff
         ssh = {
           enable = true;
-          hostRSAKey = ./secrets/dropbear_host_rsa_key;
+          hostKeys = [ "/etc/secrets-initrd/ssh_host_ed25519_key" ];
           port = 36;
-          authorizedKeys = [ "ssh-rsa (...)" ];
+          authorizedKeys = [ "ssh-ed25519 (...)" ];
+          extraConfig = ''
+            PermitRootLogin without-password
+
+            KexAlgorithms curve25519-sha256,curve25519-sha256@libssh.org
+            Ciphers chacha20-poly1305@openssh.com
+            MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com
+
+            ClientAliveInterval 30
+            ClientAliveCountMax 6
+          '';
```

The ed25519 key *does* exist on the target machine, as per the instructions related to the Dropbear -> OpenSSH initrd change.

**Expected behavior**
The new configuration should have been deployed without any errors. It should not have broken because of an unimportant detail of a previously deployed (and working) configuration!

**Screenshots**
See above.

**Additional context**

This appears related to #85418, #84976 and #85563, possibly also #82603.

In #85418, @lopsided98 mentions:
> The initrds for all generations should be generated each time you activate a new configuration.

Is there an explanation why initrd generation was designed this way? As far as I can see, the previous configurations (including initrds) should not be touched when activating new configurations, otherwise there is a risk of breaking rollbacks!

**Notify maintainers**
@lopsided98 @emilazy 

**Metadata**
> Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Deploying machine:
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.72, NixOS, 20.09.git.4d3d4221cad (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - nixpkgs: `/nixdev/nixpkgs`
```
Target machine:
```
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels/nixos' does not exist, ignoring
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I), at (string):1:9
 - system: `0`
 - host os: `Linux 5.4.70, NixOS, 20.03.git.e148d43a5f8 (Markhor)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `""`
 - channels(wizeman): `""`
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels/nixos' does not exist, ignoring
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I), at (string):1:1
 - nixpkgs: `0`
```

**Comment by wizeman** [comment]
As a workaround, I mounted the Nix store read-write and deleted all the `append-initrd-secrets` symlinks present in my 20.03 generations. Afterwards, I was able to `nixops deploy` the new 20.09 configuration and boot from it. Note, however, that I didn't test whether my 20.03 generations remained bootable or not.

**Comment by Sepero** [comment]
I'm also suffering from this bug. I followed the Remote LUKS Unlocking tutorial and ended up with this.

    cp: cannot stat '/etc/secrets/initrd/ssh_host_rsa_key': No such file or directory

I have since entirely removed all nix configuration code related to the tutorial, and the problem persists. Even `nix-rebuild switch --rollback` does not fix the issue.

https://nixos.wiki/wiki/Remote_LUKS_Unlocking

**Comment by Sepero** [comment]
> As a workaround, I mounted the Nix store read-write and deleted all the `append-initrd-secrets` symlinks present in my 20.03 generations. Afterwards, I was able to `nixops deploy` the new 20.09 configuration and boot from it. Note, however, that I didn't test whether my 20.03 generations remained bootable or not.

Following this, I was able to rebuild again

**Comment by mcdonc** [comment]
I ran into the same symptom today with a minor difference:  I wasn't upgrading between versions.

On a relatively-freshly-created 20.05 system,  I attempted to activate remote ZFS boot unlock as described at https://nixos.wiki/wiki/ZFS#Optional_additional_setup_for_encrypted_ZFS .  I couldn't make the networking bits work, so I removed the related config and the key I generated, and attempted to rebuild.

Now, although no mention of ``ssh_preboot_key`` exists in my config, each time I rebuild, I get:

    building Nix...
    building the system configuration...
    updating GRUB 2 menu...
    cp: cannot stat '/home/chrism/.nixconfig/thinknix51/ssh_preboot_key': No such file or directory
    failed to create initrd secrets: No such file or directory
    warning: error(s) occurred while switching to the new configuration

I'm not sure how to mount the nix store as read-write to try the above workaround, still searching around for info about that.

**Comment by turlando** [comment]
I am affected by this bug and it is still showing up in 22.11.

I performed a fresh install of NixOS 22.11, storing my initrd ssh keys in some path, then changing such path in a second moment. The `cannot stat '<path to old ssh key>': No such file or directory` error showed up while rebuilding. As suggested I manually removed the append-initrd-secrets files from the store and garbage collected multiple times until convergence. I have now updated the configuration again (if I didn't break the configuration the initrd ssh keys path shouldn't have changed, although I can't really guarantee it) and I'm facing this error:

```
% nixos-rebuild -I nixos-config=configuration.nix switch
building Nix...
building the system configuration...
these 8 derivations will be built:
  /nix/store/58w0gz6qq6mb4nx4jknvi9zvd1fayf0c-sshd.conf-validated.drv
  /nix/store/d1yjd3l14wjp2k8mqk5rcccz5iw6ja9k-unit-console-getty.service-disabled.drv
  /nix/store/7h06dmgk5gwwka9s8nswaxgndl1b3sf6-unit-script-sshd-pre-start.drv
  /nix/store/kvkndba0631diicv146fw34mxjkcv608-unit-sshd.service.drv
  /nix/store/63996nv84fijajgrq4m0khlcq8vxinac-system-units.drv
  /nix/store/6fnhmydnwlkf9cj58cs4vawpl3kc2rav-etc.drv
  /nix/store/rhdi87zhxq27pbnpbqgcdq88nkhb69gf-stage-2-init.sh.drv
  /nix/store/sc7q6qym66d6xdkcrqxyhfzpggnxmhfr-nixos-system-antigone-22.11.1777.cdead16a444.drv
these 5 paths will be fetched (0.70 MiB download, 9.48 MiB unpacked):
  /nix/store/0vxy2dmivgnng1f7c38q80ypkyl6zn65-gnumake-4.3
  /nix/store/6ys7v7zc0qdlnpzgj23dp8arlp90zb6q-stdenv-linux
  /nix/store/mmp8l7hxb36b9a349mpnvdblmbw552l1-lndir-1.0.3
  /nix/store/plangzrapffqqbrb7qrbvk905mdfagl5-file-5.43
  /nix/store/qfzkrv19az50ypgrsgv6ywr3pdw4rm5y-patchelf-0.15.0
error: opening file '/nix/store/l8pv1csxnsfpczj3h7azh2h5fp2jhrwa-append-initrd-secrets.drv': No such file or directory
```

`/nix/store/l8pv1csxnsfpczj3h7azh2h5fp2jhrwa-append-initrd-secrets.drv` is indeed one of the files I previously removed.

I think I will just reinstall.

Let me know if I can do something to track down this bug. It is very annoying.

**Comment by ncfavier** [comment]
I think that was fixed by [`eccc1e5` (#210812)](https://github.com/NixOS/nixpkgs/pull/210812/commits/eccc1e5bf482491187e914a4c37ba45a5de56703)?
